File: mid/Dynamic Programming/238. Product of Array Except Self.py
# https://leetcode.com/problems/product-of-array-except-self/solutions/4876857/simple-easy-approach-prefixsum-beats-73-94-c-java-kotlin/
# Time complexity: O(n)
# Space complexity: O(n)

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left = [1] * n
        right = [1] * n
        
        # Calculate left prefix product
        for i in range(1, n): # left[0] = 1
            left[i] = left[i - 1] * nums[i - 1]

        # Calculate right suffix product
        for i in range(n - 2, -1, -1):
            right[i] = right[i + 1] * nums[i + 1]

        # Calculate result array
        result = [left[i] * right[i] for i in range(n)]
       
        return result

File: mid/Dynamic Programming/238. Product of Array Except Self.cpp
// https://leetcode.com/problems/product-of-array-except-self/solutions/4876857/simple-easy-approach-prefixsum-beats-73-94-c-java-kotlin/
// Time complexity: O(n)
// Space complexity: O(1)

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int>left(n,1);
        vector<int>right(n,1);
        left[0] = 1;
        for(int i=1; i<n; i++){
            left[i] = left[i-1]*nums[i-1];
        }

        right[n-1] = 1;
        for(int i=n-2; i>=0; i--){
           right[i] = right[i+1]*nums[i+1];
        }
       vector<int>result(n);
       for(int i=0; i<n; i++){
           result[i] = left[i]*right[i];
       }
       
        return result;
    }
};